### PR TITLE
ENH: Implements wavelet denoising (from #1833)

### PR DESCRIPTION
**Pull request by stsievert**
## Description
This implements wavelet denoising and is based off #1833 -- see that PR for more detail. I tried to `git rebase` but my attempt failed and this is faster.

New changes that are not reflected in #1833:

* pywt added to optional requirements
* a parameter renamed (`noise_std` --> `sigma`).

## Checklist
[It's fine to submit PRs which are a work in progress! But before they are merged, all PRs should provide:]
- [x] Clean style in [the spirit of PEP8](https://www.python.org/dev/peps/pep-0008/)
- [x] [Docstrings for all functions](https://github.com/numpy/numpy/blob/master/doc/example.py)
- [x] Gallery example in `./doc/examples` (new features only)
- [x] Unit tests
- [x] Add to contributors.txt

## References
This is the same implementation as #1833, but I made a new branch to resolve some merge conflicts (I tried to rebase but *sigh*).

### CODE REVIEW

File: skimage/restoration/_denoise.py [hunk 1]
```diff
@@ -332,3 +332,126 @@ def denoise_tv_chambolle(im, weight=0.1, eps=2.e-4, n_iter_max=200,
...
+        # The BayesShrink threshold from [1]_ in docstring
+        threshold = sigma**2 / np.sqrt(max(im.var() - sigma**2, 0))
+
+    denoised_detail = [{key: pywt.threshold(level[key], value=threshold,
+                       mode=mode) for key in level} for level in coeffs[1:]]
+    denoised_root = pywt.threshold(coeffs[0], value=threshold, mode=mode)
+    return pywt.waverecn([denoised_root, *denoised_detail], wavelet)
+
+
+def denoise_wavelet(im, sigma=None, wavelet='db1', mode='soft'):
```
Comment (sciunto) [inline skimage/restoration/_denoise.py]: We prefer image (or img) instead of im.

File: skimage/restoration/_denoise.py [hunk 2]
```diff
@@ -332,3 +332,126 @@ def denoise_tv_chambolle(im, weight=0.1, eps=2.e-4, n_iter_max=200,
     else:
         out = _denoise_tv_chambolle_nd(im, weight, eps, n_iter_max)
     return out
+
+
+def _wavelet_threshold(im, wavelet, threshold=None, sigma=None, mode='soft'):
```
Comment (sciunto) [inline skimage/restoration/_denoise.py]: We prefer image (or img) instead of im.

File: skimage/restoration/_denoise.py [hunk 3]
```diff
@@ -332,3 +332,126 @@ def denoise_tv_chambolle(im, weight=0.1, eps=2.e-4, n_iter_max=200,
...
+    """Performs wavelet denoising.
```
Comment (sciunto) [inline skimage/restoration/_denoise.py]: Please, add a blank line

File: skimage/restoration/_denoise.py [hunk 4]
```diff
@@ -332,3 +332,126 @@ def denoise_tv_chambolle(im, weight=0.1, eps=2.e-4, n_iter_max=200,
...
+        threshold = sigma**2 / np.sqrt(max(im.var() - sigma**2, 0))
+
+    denoised_detail = [{key: pywt.threshold(level[key], value=threshold,
+                       mode=mode) for key in level} for level in coeffs[1:]]
+    denoised_root = pywt.threshold(coeffs[0], value=threshold, mode=mode)
+    return pywt.waverecn([denoised_root, *denoised_detail], wavelet)
+
+
+def denoise_wavelet(im, sigma=None, wavelet='db1', mode='soft'):
+    """Performs wavelet denoising on an image.
```
Comment (sciunto) [inline skimage/restoration/_denoise.py]: Please, add a blank line

File: skimage/restoration/_denoise.py [hunk 5]
```diff
@@ -332,3 +332,126 @@ def denoise_tv_chambolle(im, weight=0.1, eps=2.e-4, n_iter_max=200,
...
+        The noise standard deviation used when computing the threshold
+        adaptively as described in [1]_.
+    wavelet : string, optional
+        The type of wavelet to perform and can be any of the options
+        [pywt.wavelist]_ outputs. The default is `'db1'`. For example,
+        ``wavelet`` can be any of ``{'db2', 'haar', 'sym9'}`` and many more.
+    mode : {'soft', 'hard'}, optional
+        An optional argument to choose the type of denoising performed. It
+        noted that choosing soft thresholding given additive noise finds the
+        best approximation of the original image.
```
Comment (sciunto) [inline skimage/restoration/_denoise.py]: Blank line

File: skimage/restoration/_denoise.py [hunk 6]
```diff
@@ -332,3 +332,126 @@ def denoise_tv_chambolle(im, weight=0.1, eps=2.e-4, n_iter_max=200,
...
+    Returns
+    -------
+    out : ndarray
+        Denoised image.
```
Comment (sciunto) [inline skimage/restoration/_denoise.py]: also here. blank line between section headers

File: skimage/restoration/_denoise.py [hunk 7]
```diff
@@ -332,3 +332,126 @@ def denoise_tv_chambolle(im, weight=0.1, eps=2.e-4, n_iter_max=200,
...
+    wavelet domain. Correspondingly, many pixel values of an image are 0 after
+    taking the wavelet transform.
+    By wavelet denoising, we are enforcing that many of the wavelet coefficients
+    are 0 while keeping the error small. When we use soft thresholding, our
+    estimate is
+    .. math:: \widehat{x} = \arg \min_x ||z - x||_2^2 + \lambda ||x||_1
+    where :math:`z` is the input image wavelet coefficients and :math:`\lambda`
+    is the threshold.
+    This function performs wavelet denoising on each color plane separately. The
+    output is clipped between 0 and 1.
```
Comment (sciunto) [inline skimage/restoration/_denoise.py]: same here (blank line)

File: skimage/restoration/_denoise.py [hunk 8]
```diff
@@ -332,3 +332,126 @@ def denoise_tv_chambolle(im, weight=0.1, eps=2.e-4, n_iter_max=200,
...
+    ----------
+    .. [1] Chang, S. Grace, Bin Yu, and Martin Vetterli. "Adaptive wavelet
+           thresholding for image denoising and compression." Image Processing,
+           IEEE Transactions on 9.9 (2000): 1532-1546.
+    .. [pywt.wavelist] http://pywavelets.readthedocs.org/en/latest/ref/wavelets.html#wavelet-wavelist
+
+    Examples
+    --------
+    >>> from skimage import color, data
+    >>> img = data.astronaut() * 1.0 / 255
```
Comment (sciunto) [inline skimage/restoration/_denoise.py]: Why not img_as_float()?

**Comment by sciunto** [comment]
You need to add your function to `skimage/restauration/__init__.py`

File: optional_requirements.txt [hunk 9]
```diff
@@ -4,3 +4,4 @@ SimpleITK; python_version != '3.4' and python_version != '3.5'
 astropy
 tifffile
 imageio
+PyWavelet
```
Comment (sciunto) [inline optional_requirements.txt]: I don't think it's an optional requirement. Should go in requirements.txt. That's why travis didn't install it.

**Comment by sciunto** [comment]
@stsievert I made a first pass and wrote some comments to give you some indications to improve different things and to pass the tests.

It's good to see wavelets coming in :)

File: skimage/restoration/_denoise.py [hunk 10]
```diff
@@ -332,3 +332,126 @@ def denoise_tv_chambolle(im, weight=0.1, eps=2.e-4, n_iter_max=200,
...
+    Returns
+    -------
+    out : ndarray
+        Denoised image.
+
+    References
+    ----------
+    .. [1] Chang, S. Grace, Bin Yu, and Martin Vetterli. "Adaptive wavelet
+           thresholding for image denoising and compression." Image Processing,
+           IEEE Transactions on 9.9 (2000): 1532-1546.
```
Comment (sciunto) [inline skimage/restoration/_denoise.py]: Please, add DOI. Format: `DOI:xxxyyy`

File: skimage/restoration/_denoise.py [hunk 11]
```diff
@@ -332,3 +332,126 @@ def denoise_tv_chambolle(im, weight=0.1, eps=2.e-4, n_iter_max=200,
...
+    .. math:: \widehat{x} = \arg \min_x ||z - x||_2^2 + \lambda ||x||_1
+    where :math:`z` is the input image wavelet coefficients and :math:`\lambda`
+    is the threshold.
+    This function performs wavelet denoising on each color plane separately. The
+    output is clipped between 0 and 1.
+    References
+    ----------
+    .. [1] Chang, S. Grace, Bin Yu, and Martin Vetterli. "Adaptive wavelet
+           thresholding for image denoising and compression." Image Processing,
+           IEEE Transactions on 9.9 (2000): 1532-1546.
```
Comment (sciunto) [inline skimage/restoration/_denoise.py]: Please, add DOI.

File: skimage/restoration/tests/test_denoise.py [hunk 12]
```diff
@@ -308,5 +308,21 @@ def test_no_denoising_for_small_h():
     assert np.allclose(denoised, img)
 
 
+def test_wavelet_denoising():
```
Comment (sciunto) [inline skimage/restoration/tests/test_denoise.py]: I think we need at least two tests. One for 2d, one for 3d.

File: skimage/restoration/_denoise.py [hunk 13]
```diff
@@ -332,3 +332,126 @@ def denoise_tv_chambolle(im, weight=0.1, eps=2.e-4, n_iter_max=200,
...
+    """
+    import pywt
+    coeffs = pywt.wavedecn(im, wavelet=wavelet)
+    detail_coeffs = coeffs[-1]['d' * im.ndim]
+
+    if sigma is None:
+        # Estimate the noise std.dev as discussed in PR #1837
```
Comment (sciunto) [inline skimage/restoration/_denoise.py]: I prefer to avoid references to PR as much as possible. Could you summarize the idea please?

File: skimage/restoration/_denoise.py [hunk 14]
```diff
@@ -332,3 +332,142 @@ def denoise_tv_chambolle(im, weight=0.1, eps=2.e-4, n_iter_max=200,
...
+    ----------
+    .. [1] Chang, S. Grace, Bin Yu, and Martin Vetterli. "Adaptive wavelet
+           thresholding for image denoising and compression." Image Processing,
+           IEEE Transactions on 9.9 (2000): 1532-1546.
+           DOI: 10.1109/83.862633
+    .. [2] D. L. Donoho and I. M. Johnstone. "Ideal spatial adaptation
+           by wavelet shrinkage." Biometrika 81.3 (1994): 425-455.
+           DOI: 10.1093/biomet/81.3.425
+    """
+    import pywt
```
Comment (sciunto) [inline skimage/restoration/_denoise.py]: Why do you import pywt in the function?

Comment (stsievert) [inline skimage/restoration/_denoise.py]: I'm trying to keep it an optional requirement -- that's what had been decided a couple months ago.

File: skimage/restoration/_denoise.py [hunk 15]
```diff
@@ -332,3 +332,142 @@ def denoise_tv_chambolle(im, weight=0.1, eps=2.e-4, n_iter_max=200,
...
+        sigma = np.median(np.abs(detail_coeffs)) / 0.67448975019608171
+
+    if threshold is None:
+        # The BayesShrink threshold from [1]_ in docstring
+        threshold = sigma**2 / np.sqrt(max(img.var() - sigma**2, 0))
+
+    denoised_detail = [{key: pywt.threshold(level[key], value=threshold,
+                       mode=mode) for key in level} for level in coeffs[1:]]
+    denoised_root = pywt.threshold(coeffs[0], value=threshold, mode=mode)
+    return pywt.waverecn([denoised_root, *denoised_detail], wavelet)
```
Comment (sciunto) [inline skimage/restoration/_denoise.py]: splat in a list is python 3 (3.5?) specific.

File: skimage/restoration/_denoise.py (hunk 14 above)
Comment (sciunto) [inline skimage/restoration/_denoise.py]: I'm not sure that the decision is actually taken. See #1838.

Comment (soupault) [inline skimage/restoration/_denoise.py]: @sciunto we have actually reached an agreement on this in https://github.com/scikit-image/scikit-image/pull/2019 .

**Comment by stsievert** [comment]
I've made PyWavelets required. It seems that in #2019 we decided that we would wait for a PR of this flavor to come along.

**Comment by sciunto** [comment]
I'm +1 to do it this way for now and eventually turn off before the release if, for some reasons, we want to wait a little.

@stsievert a little indent issue, then, we should be close :)

**Comment by stsievert** [comment]
I have resolved the indentation issue.

The Travis CI OS X build is still failing and didn't generate a log. Sounds like the Travis CI build needs to be restarted?

**Comment by sciunto** [comment]
@stsievert Looks running from here. Sometimes, this one fails unexpectedly. If we close and reopen the PR, we can trigger a new run.

File: doc/examples/filters/plot_denoise.py [hunk 16]
```diff
@@ -38,7 +38,7 @@
 noisy = astro + 0.6 * astro.std() * np.random.random(astro.shape)
```
Comment (sciunto) [inline doc/examples/filters/plot_denoise.py]: I'm wondering if using http://scikit-image.org/docs/dev/api/skimage.util.html#random-noise wouldn't be better... @jni just fixed this function, what's your opinion @jni?

File: skimage/restoration/_denoise.py [hunk 17]
```diff
@@ -332,3 +333,135 @@ def denoise_tv_chambolle(im, weight=0.1, eps=2.e-4, n_iter_max=200,
...
     return out
+
+
+def _wavelet_threshold(img, wavelet, threshold=None, sigma=None, mode='soft'):
+    """Performs wavelet denoising.
+
+    Parameters
+    ----------
+    img : ndarray (2d or 3d) of ints, uints or floats
+        Input data to be denoised. `im` can be of any numeric type,
```
Comment (sciunto) [inline skimage/restoration/_denoise.py]: im -> img

File: skimage/restoration/_denoise.py [hunk 18]
```diff
@@ -332,3 +333,135 @@ def denoise_tv_chambolle(im, weight=0.1, eps=2.e-4, n_iter_max=200,
...
+
+    """
+
+    if not img.dtype.kind == 'f':
+        img = img_as_float(img)
+
+    if img.ndim == 2:
+        out = _wavelet_threshold(img, wavelet=wavelet, mode=mode,
+                                 sigma=sigma)
+
```
Comment (sciunto) [inline skimage/restoration/_denoise.py]: remove blank lines between if and else.

**Comment by sciunto** [comment]
I left little comments, otherwise, LGTM. We just need another pair of eyes.

File: doc/examples/filters/plot_denoise.py (hunk 16 above)
Comment (jni) [inline doc/examples/filters/plot_denoise.py]: LOL I am definitely far from the expert here, but yes, it seems like that's what those functions are for! I would assume Gaussian noise would be more realistic than uniform noise, too...

Comment (stsievert) [inline doc/examples/filters/plot_denoise.py]: @sciunto resolved. Instead of uniform noise I use Gaussian nose through `random_noise` with the same standard deviation.

**Comment by sciunto** [comment]
Perfect! LGTM :)

File: skimage/restoration/_denoise.py [hunk 19]
```diff
@@ -332,3 +333,134 @@ def denoise_tv_chambolle(im, weight=0.1, eps=2.e-4, n_iter_max=200,
...
+    >>> from skimage import color, data
+    >>> img = img_as_float(data.astronaut())
+    >>> img = color.rgb2gray(img)
+    >>> img += 0.1 * np.random.randn(*img.shape)
+    >>> img = np.clip(img, 0, 1)
+    >>> denoised_img = denoise_wavelet(img, sigma=0.1)
+
+    """
+
+    if not img.dtype.kind == 'f':
```
Comment (ahojnnes) [inline skimage/restoration/_denoise.py]: This if condition could be removed. img_as_float will only do something if the image is not already float.

**Comment by ahojnnes** [comment]
One small comment and Travis is not happy, otherwise LGTM!

**Comment by JDWarner** [comment]
Travis is only unhappy on OSX, where it seems the build isn't even trying to start.

**Comment by sciunto** [comment]
and you need to add your name to the contributor list :)

File: skimage/restoration/_denoise.py [hunk 20]
```diff
@@ -332,3 +333,134 @@ def denoise_tv_chambolle(im, weight=0.1, eps=2.e-4, n_iter_max=200,
...
+    denoised_coeffs = [denoised_root] + [d for d in denoised_detail]
+    return pywt.waverecn(denoised_coeffs, wavelet)
+
+
+def denoise_wavelet(img, sigma=None, wavelet='db1', mode='soft'):
+    """Performs wavelet denoising on an image.
+
+    Parameters
+    ----------
+    img : ndarray (greater than 2d) of ints, uints or floats
```
Comment (soupault) [inline skimage/restoration/_denoise.py]: >=2D here, but 2D/3D only in the function above.
Could you clarify the supported formats and fix the docstrings accordingly?

File: skimage/restoration/_denoise.py [hunk 21]
```diff
@@ -332,3 +333,134 @@ def denoise_tv_chambolle(im, weight=0.1, eps=2.e-4, n_iter_max=200,
...
+    if img.ndim == 2:
+        out = _wavelet_threshold(img, wavelet=wavelet, mode=mode,
+                                 sigma=sigma)
+    else:
+        out = np.dstack([_wavelet_threshold(img[..., c], wavelet=wavelet,
+                                            mode=mode, sigma=sigma)
+                         for c in range(img.ndim)])
+
+    # ensure valid image in 0, 1 is returned
+    return np.clip(out, 0, 1)
```
Comment (soupault) [inline skimage/restoration/_denoise.py]: Not correct if the image range is [-1; 1].
Should check the range at the beginning and clip to the one being used.

File: skimage/restoration/tests/test_denoise.py [hunk 22]
```diff
@@ -308,5 +308,21 @@ def test_no_denoising_for_small_h():
...
+        img = np.clip(img, 0, 1)
+        # less energy in signal
+        assert restoration.denoise_wavelet(img, sigma=0.3).sum() <= img.sum()
+
+        # test changing noise_std (higher threshold, so less energy in signal)
+        assert (restoration.denoise_wavelet(img, sigma=0.2).sum() <=
+                restoration.denoise_wavelet(img, sigma=0.1).sum())
+
+        # This works for this particular image
+        assert (np.sum(restoration.denoise_wavelet(img, wavelet='db2')) <
```
Comment (soupault) [inline skimage/restoration/tests/test_denoise.py]: I'd remove this case as it doesn't add (IMO) any value to the testing.

File: skimage/restoration/_denoise.py [hunk 23]
```diff
@@ -332,3 +333,134 @@ def denoise_tv_chambolle(im, weight=0.1, eps=2.e-4, n_iter_max=200,
...
+        best approximation of the original image.
+
+    Returns
+    -------
+    out : ndarray
+        Denoised image.
+
+    Notes
+    -----
+    As with the Fourier transform, there is an analogue to frequency in the
```
Comment (JDWarner) [inline skimage/restoration/_denoise.py]: I might say this a little differently. Something like

"The wavelet domain is a sparse representation of the image, and can be thought of similarly to the frequency domain of the Fourier transform. Sparse representations have most values zero or near-zero. Truly random noise is usually represented by many small values in the wavelet domain. Thus, setting values in the wavelet domain below some threshold to zero reduces noise. Higher thresholds remove more noise, but will reduce fine detail."

File: skimage/restoration/tests/test_denoise.py (hunk 22 above)
Comment (JDWarner) [inline skimage/restoration/tests/test_denoise.py]: I think a good test would be comparing the original image vs. noisy with `compare_ssim` and the original vs. denoised also with `compare_ssim`. The denoised comparison should be better.

**Comment by JDWarner** [comment]
I agree with @soupault's comments. Once addressed, this should be good to go.

File: skimage/restoration/_denoise.py [hunk 24]
```diff
@@ -332,3 +333,140 @@ def denoise_tv_chambolle(im, weight=0.1, eps=2.e-4, n_iter_max=200,
...
+    >>> img = color.rgb2gray(img)
+    >>> img += 0.1 * np.random.randn(*img.shape)
+    >>> img = np.clip(img, 0, 1)
+    >>> denoised_img = denoise_wavelet(img, sigma=0.1)
+
+    """
+
+    img = img_as_float(img)
+
+    if img.ndims not in {2, 3}:
```
Comment (JDWarner) [inline skimage/restoration/_denoise.py]: `img.ndim`

File: skimage/restoration/tests/test_denoise.py (hunk 22 above)
Comment (stsievert) [inline skimage/restoration/tests/test_denoise.py]: I deleted the extra test that didn't add any value. @JDWarner my test of `compare_ssim` failed (plus not familiar). The main motivation for the current test is that the energy in the space and wavelet domains is constant.

**Comment by JDWarner** [comment]
Thanks for sticking with this, @stsievert - the AppVeyor issue is unrelated, Travis is happy, and I'm merging.

**Comment by soupault** [comment]
[reminder] To review: https://github.com/scikit-image/scikit-image/issues/1838, https://github.com/scikit-image/scikit-image/issues/602, https://github.com/scikit-image/scikit-image/pull/760.

**Comment by soupault** [comment]
Hmm. I think the example is not really representative (http://scikit-image.org/docs/dev/auto_examples/filters/plot_denoise.html). Maybe apply stronger filtering?

**Comment by sciunto** [comment]
+1 and how about grayscale vs color images? Definitely not the same difficulty.

**Comment by JDWarner** [comment]
It's actually significantly denoised, but doing the denoising separately per channel is problematic - each gets handled differently, so you end actually adding color noise.

**Comment by JDWarner** [comment]
A common approach is to denoise the Y channel in YCbCr space, then return to RGB.

**Comment by stsievert** [comment]
Looks like in the top image we specify that the noise variation is less than it actually is. I say change the top image performs as default and the bottom image reduces more noise.

At the very least we should add a note about YCbCr colorspace in the docs, and we should probably show the output when denoising using the YCbCr colorspace.

**Comment by sciunto** [comment]
@stsievert Would you like to start a PR on YCbCr? I would be appreciated for the next release :)

**Comment by stsievert** [comment]
I am interested in doing that, but am busy for the next week.
